test/Test.hs

```diff
@@ -1,7 +1,191 @@
+  s <- genSeqTSized m (sz - 1)
+  pure (a :< s)
+
+genSeqTSized :: forall m a. MonadGen m => m a -> Size -> m (SeqT TestM a)
+genSeqTSized m sz = do
+  part <- splat sz
+  goop <- traverse (genTestMSized (genViewSized m)) part
+  pure $ SeqT $ listToQueue goop
+
+genSeqT :: forall m a. MonadGen m => m a -> m (SeqT TestM a)
+genSeqT m = Gen.sized $ \sz -> do
+  tsz <- Gen.integral (Range.linear 0 sz)
+  genSeqTSized m tsz
+  
+simpleSeqT :: MonadGen m => m (SeqT TestM Int)
+simpleSeqT = genSeqT (Gen.integral $ Range.constant 0 5)
+
+simpleSeqT' :: MonadGen m => m (Size, SeqT TestM Int)
+simpleSeqT' = Gen.sized $ \s -> do
+  res <- Gen.resize s $ genSeqT (Gen.integral $ Range.constant 0 5)
+  pure (s, res)
 
 main :: IO ()
 main = hspec $ do
-  return ()
+  describe "observe" $ do
+    it "undoes pure" $ hedgehog $
+      observe (pure (3 :: Int)) === 3
+  describe "observeT" $ do
+    it "undoes lift" $ hedgehog $ do
+      ex <- forAll simpleTestM
+      runMaybeT (observeT (lift (lift ex))) === runMaybeT (lift ex)
+  describe "observeAllT" $ do
+    it "undoes lift" $ hedgehog $ do
+      ex <- forAll simpleTestM
+      observeAllT (lift ex) === fmap (:[]) ex
+    it "works like logicT" $ hedgehog $ do
+      ex <- forAll simpleSeqT
+      observeAllT ex === L.observeAllT (fromSeqT ex)
+  describe ">>=" $ do
+    it "obeys monad identity law 1" $ hedgehog $ do
+      s <- forAll simpleSeqT
+      (s >>= return) === s
+    it "obeys monad identity law 2" $ hedgehog $ do
+      a <- forAll $ Gen.integral Range.linearBounded
+      f :: Int -> SeqT TestM Int <- Fun.forAllFn (Fun.fn simpleSeqT)
+      (pure a >>= f) === f a
+    it "works like LogicT" $ hedgehog $ do
+      s <- forAll simpleSeqT
+      f :: Int -> SeqT TestM Int <- Fun.forAllFn (Fun.fn simpleSeqT)
+      fromLogicT (toLogicT s >>= toLogicT . f) === (s >>= f)
+    it "obeys monad associativity law" $ hedgehog $ do
+      s <- forAll simpleSeqT
+      f :: Int -> SeqT TestM Int <- Fun.forAllFn (Fun.fn simpleSeqT)
+      g :: Int -> SeqT TestM Int <- Fun.forAllFn (Fun.fn simpleSeqT)
+      ((s >>= f) >>= g) === (s >>= \a -> f a >>= g)
+  describe "fromLogicT" $ do
+    it "reverses fromSeqT" $ hedgehog $ do
+      s <- forAll simpleSeqT
+      fromLogicT (fromSeqT s) === s
+  describe "fromView" $ do
+    it "reverses toView" $ hedgehog $ do
+      s <- forAll simpleSeqT
+      fromView (toView s) === s
+  describe "MonadReader instance" $ do
+    it "passes the tests in https://github.com/Bodigrim/logict/issues/1" $ do
+      runReader (runMaybeT (observeAllT (local (5+) ask))) 0 `shouldBe` Just [5]
+      let
+        foo :: MonadReader Int m => m (Int, Int)
+        foo = do
+          x <- local (5+) ask
+          y <- ask
+          return (x, y)
+      runReader (runMaybeT (observeT foo)) 0 `shouldBe` Just (5, 0)
+  describe "MFunctor instance" $ do
+    it "obeys the hoist identity law" $ hedgehog $ do
+      s <- forAll simpleSeqT
+      hoist (\x -> x) s === s
+
+{-
```

Comment:
Is there any reason to keep this commented code? My preference would be to deleted it.

Reply:
Yeah, I can do that.